server/chat-plugins/logviewer.ts

```diff
@@ -0,0 +1,138 @@
+import {FS} from '../../lib/fs';
+
+interface RoomlogReader {
+	readLogs: (roomid: RoomID, year: string, month: string, date: string) => string;
+	directory: (room: Room) => string;
+}
+class RoomlogReaderFS implements RoomlogReader {
+	readLogs(roomid: RoomID, year: string, month: string, date: string) {
+		 let buf;
+		try {
+			buf = FS(`logs/chat/${roomid}/${year}-${month}/${year}-${month}-${date}.txt`).readIfExistsSync();
+		} catch (e) {
+			buf = '';
+		}
+		return buf;
+	}
+
+	directory(room: Room) {
+		let buf = "&nbsp;&nbsp;<h2><b>Select a date to view logs.</b></h2>";
+		const months = FS(`logs/chat/${room}/`).readdirSync();
+		for (const m of months) {
+			if (!FS(`logs/chat/${room}/${m}`).isDirectorySync()) continue;
+			buf += `<br><br><b><center>${m}</center></b><br>`;
+			for (const file of FS(`logs/chat/${room}/${m}`).readdirSync()) {
+				if (!file.endsWith('.txt')) continue;
+				buf += `${this.button(file.slice(0, -4), room)}`;
+			}
+		}
+
+		return buf;
+	}
+
+	month(month: string, room: Room) {
+		let buf = "&nbsp;&nbsp;<h2><b>Select a date to view logs.</b></h2>";
+		const days = FS(`logs/chat/${room}/${month}`).readdirSync();
+		for (const file of days) {
+			if (!file.endsWith('.txt')) continue;
+			buf += `${this.button(file.slice(0, -4), room)}`;
+		}
+
+		return buf;
+	}
+	
+	button(datestring: string, room: Room, nodate = false) {
```

Comment:
This has no relation to reading, so it should be in `RoomlogViewer`.

Reply:
directory() needs it, so i shifted it because i wasn't sure if it was possible to call from there

Reply:
I think a better solution is to make `RoomlogReader#directory` return an array of roomids and then `RoomlogViewer#directory` will turn it to HTML, so then the button can be moved back to `RoomlogViewer`.

Reply:
Maybe month should have the same refactor done on it to have a clean `RoomlogReader` API (ie. not intermingled with HTML).

Reply:
yea i'm working on that right now